src/asm_pattern.py

```diff
@@ -0,0 +1,182 @@
+        else:
+            assert False, f"bad pattern part in math pattern: {e}"
+
+    def match_one(self, actual: BodyPart, exp: PatternPart) -> bool:
+        if exp is None:
+            return True
+        if isinstance(exp, Label):
+            name = self.symbolic_labels.get(exp.name)
+            return isinstance(actual, Label) and (
+                name is None or actual.name == name
+            )
+        if not isinstance(actual, Instruction):
+            return False
+        ins = actual
+        if ins.mnemonic != exp.mnemonic:
+            return False
+        if exp.args:
+            if len(exp.args) != len(ins.args):
+                return False
+            for (e, a) in zip(exp.args, ins.args):
+                if isinstance(e, AsmLiteral):
+                    if not isinstance(a, AsmLiteral) or a.value != e.value:
+                        return False
+                elif isinstance(e, Register):
+                    if not isinstance(a, Register) or not self.match_reg(a, e):
+                        return False
+                elif isinstance(e, AsmGlobalSymbol):
+                    if e.symbol_name.isupper():
+                        if not isinstance(a, AsmLiteral):
+                            return False
+                        if e.symbol_name not in self.symbolic_literals:
+                            self.symbolic_literals[e.symbol_name] = a.value
+                        elif self.symbolic_literals[e.symbol_name] != a.value:
+                            return False
+                    else:
+                        if (
+                            not isinstance(a, AsmGlobalSymbol)
+                            or a.symbol_name != e.symbol_name
+                        ):
+                            return False
+                elif isinstance(e, AsmAddressMode):
+                    if (
+                        not isinstance(a, AsmAddressMode)
+                        or a.lhs != e.lhs
+                        or not self.match_reg(a.rhs, e.rhs)
+                    ):
+                        return False
+                elif isinstance(e, JumpTarget):
+                    if not isinstance(a, JumpTarget):
+                        return False
+                    if e.target not in self.symbolic_labels:
+                        self.symbolic_labels[e.target] = a.target
+                    elif self.symbolic_labels[e.target] != a.target:
+                        return False
+                elif isinstance(e, BinOp):
+                    if not isinstance(a, AsmLiteral) or a.value != self.eval_math(e):
+                        return False
+                else:
+                    assert False, f"bad pattern part: {exp} contains {type(e)}"
+        return True
+
+
+@dataclass
+class AsmMatcher:
+    input: List[BodyPart]
+    output: List[BodyPart] = field(default_factory=list)
+    index: int = 0
+
+    def try_match(self, pattern: Pattern) -> Optional[List[BodyPart]]:
```

Comment:
It might be nice to return a new class for the results, instead of `List[BodyPart]`? 
If there were methods like `.get_instruction(index: int) -> Instruction`, this could remove the `typing.cast` calls? (They would be `assert`s in the getters instead.)

(I've also wanted to the symbolic registers/literals from the `TryMatchState`? The replacement functions could be easier to read with `match.get_register("x")` instead of `match[0].args[1]`...)

Reply:
Agreed! I'll do that